src/core/import-creds.c

```diff
@@ -0,0 +1,551 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <sys/mount.h>
+
+#include "copy.h"
+#include "creds-util.h"
+#include "fileio.h"
+#include "format-util.h"
+#include "fs-util.h"
+#include "import-creds.h"
+#include "io-util.h"
+#include "mkdir-label.h"
+#include "mount-util.h"
+#include "mountpoint-util.h"
+#include "parse-util.h"
+#include "path-util.h"
+#include "proc-cmdline.h"
+#include "recurse-dir.h"
+#include "strv.h"
+
+/* This imports credentials passed in from environments higher up (VM manager, boot loader, …) and rearranges
+ * them so that later code can access them using our regular credential protocol
+ * (i.e. $CREDENTIALS_DIRECTORY). It's supposed to be minimal glue to unify behaviour how PID 1 (and
+ * generators invoked by it) can acquire credentials from outside, to mimic how we support it for containers,
+ * but on VM/physical environments.
+ *
+ * This does three things:
+ *
+ * 1. It imports credentials picked up by sd-boot (and placed in the /.extra/credentials/ dir in the initrd)
+ *    and puts them in /run/credentials/@encrypted/. Note that during the initrd→host transition the initrd root
+ *    file system is cleaned out, thus it is essential we pick up these files before they are deleted. Note
+ *    that these credentials originate from an untrusted source, i.e. the ESP and are not
+ *    pre-authenticated. They still have to be authenticated before use.
```

Comment:
I'm not sure I follow this bit. If I understand correctly, this directory is intended for use with `LoadCredentialEncrypted`. Does that mean all credentials that end up in this directory must be encrypted with `sd-creds encrypt`? And does that mean systemd still handles all that for the user before passing the credentials to a service?

Reply:
well it kinda means that, but also not.

It's more about ensuring that automatic propagation doesn't happen for stuff we can't trust because it comes from an untrusted source. i.e. we should not auto-propagate stuff into services if it potentially contains rogue stuff. 

What do I mean by "auto-propagate"? mostly if LoadCredential= is used with just a cred name (ie. without specifying an explicit source), so that the credential is searched automatically in /etc/credstore/ or among the system creds. That said, it's actual OK if we propagate such an untrusted cred if the unit file makes this explicit. How can it make it explicit? By specifying a full, absolute source path. Or in other words:

A simple line like this:
```
LoadCredential=foo
```
should auto-search the credential to propagate from stuff in /etc/credstore/ and among creds passed in from a trusted hypervisor/container manager or so.

vs.

But for loading untrusted creds to pass to a service, we should require the developer to explicitly indicate that it's OK to source from an unencrypted (i.e. untrusted) source, via a line like this:

```
LoadCredential=foo:/run/credentials/@encrypted/foo
```

This line will basically upgrade the untrusted system credential "foo" into a trusted  service credential "foo" *without* validation.

A much better way of course is this:

```
LoadCredentialEncrypted=foo
```
 
This will auto-search and then validate/decrypt the credential and pass it on. This line hence takes an untrusted system credential "foo" and turns it into a trusted service credential *with* validation.

or in other words: we place trusted/untrusted credentials in two separate dirs so that the creds passed in that way are not picked up by code that expects them to be trusted even though they are not trusted at all, because they are loaded from an untrusted source (i.e. the ESP).we want to make sure that we have two distinct pools: one which contains "trusted" credentials, i.e. which can just be used, as the come from a trusted source. and "untrusted" credentials, that still need to be validated before use. It's OK if credentials in the "trusted" pool are consumed with little restrictions and hence LoadCredential= will just search the stuff from the trusted pool. However stuff from the "untrusted" pool should not consumed that easily, it either requires validation against something, or the user should explicitly say that it is OK to propagate this cred without authentication. 

Why did I call the dir "encrypted" rather than "untrusted"? Initially while I worked on it I actually named it that way. But I changed this to "encrypted" because we have so far not used the term "trust" in code for this purpose. So far we only distinguish encrpted and not encrypted creds, and the authentication is implicit to the encryption mechanism. (i.e. we actually used authenticated encryption, but this is only ever mentioned in the docs, not in the naming of the functionality). So in order to simplify things and in the hope that people will more likely use this properly I want to stick to this: there are only two types of creds, and the one way to get from one to the other is through LoadCredentialEncrypt=, and only then things are nice, and automatic.

does that make any sense?

Reply:
That makes sense, thank you for the detailed explanation.

One thing I don't completely understand is why credentials picked up by sd-boot would be untrusted. What exactly distinguishes those credentials from the trusted ones that makes them untrusted?

Reply:
> One thing I don't completely understand is why credentials picked up by sd-boot would be untrusted. What exactly distinguishes those credentials from the trusted ones that makes them untrusted?

In any environment that uses FDE or verity or something like that I think we can assume the areas of the file hierarchy that are only writable to root can be reasonably considered "trusted", as then there's protection that neither online nor offline (at least to some degree) it can be changed.

This is quite different from data stored in the ESP: it's trivially accessible offline, there is no encryption or integrity checking, it's a FAT file system you can trivially modify.

Of course, if you do not employ FDE then there's no difference in the guarantees you get for /etc/ and for the ESP but I think we should focus on secure scenarios where it is used, and thus distinguish stuff that never comes with any sort of trust guarantee from the stuff that will on systems using FDE comes with them (even if people can turn that off in which case it is not any better than the other stuff)

Reply:
Got it, I didn't think of offline modification, that makes total sense